preview/Win7Msix/Win7MSIXInstaller/ValidateTargetDeviceFamily.cpp

```diff
@@ -0,0 +1,174 @@
+    {
+        TraceLoggingWrite(g_MsixTraceLoggingProvider,
+            "Target device family name and manifest min version are compatible with OS",
+            TraceLoggingLevel(WINEVENT_LEVEL_INFO));
+    }
+    else
+    {
+        TraceLoggingWrite(g_MsixTraceLoggingProvider,
+            "Target device family name or manifest min version are not compatible with the OS",
+            TraceLoggingLevel(WINEVENT_LEVEL_ERROR));
+        return HRESULT_FROM_WIN32(ERROR_INSTALL_REJECTED);
+    }
+    return S_OK;
+}
+
+HRESULT ValidateTargetDeviceFamily::ParseTargetDeviceFamilyFromPackage()
+{
+    PackageInfo* packageInfo = m_msixRequest->GetPackageInfo();
+
+    ComPtr<IMsixDocumentElement> domElement;
+    RETURN_IF_FAILED(packageInfo->GetManifestReader()->QueryInterface(UuidOfImpl<IMsixDocumentElement>::iid, reinterpret_cast<void**>(&domElement)));
+
+    ComPtr<IMsixElement> element;
+    RETURN_IF_FAILED(domElement->GetDocumentElement(&element));
+
+    /// Obtain the TargetDeviceFamily tag info
+    ComPtr<IMsixElementEnumerator> dependencyEnum;
+    RETURN_IF_FAILED(element->GetElements(
+        L"/*[local-name()='Package']/*[local-name()='Dependencies']/*[local-name()='TargetDeviceFamily']",
+        &dependencyEnum));
+
+    BOOL hc = FALSE;
+    RETURN_IF_FAILED(dependencyEnum->GetHasCurrent(&hc));
+    if (!hc)
+    {
+        TraceLoggingWrite(g_MsixTraceLoggingProvider,
+            "No Target device family Found",
+            TraceLoggingLevel(WINEVENT_LEVEL_ERROR));
+        return HRESULT_FROM_WIN32(ERROR_INSTALL_REJECTED);
+    }
+
+    ComPtr<IMsixElement> dependencyElement;
+    RETURN_IF_FAILED(dependencyEnum->GetCurrent(&dependencyElement));
+
+    Text<wchar_t> targetDeviceFamilyName;
+    RETURN_IF_FAILED(dependencyElement->GetAttributeValue(L"Name", &targetDeviceFamilyName));
+    m_targetDeviceFamilyName = targetDeviceFamilyName.Get();
+
+    Text<wchar_t> minVersion;
+    RETURN_IF_FAILED(dependencyElement->GetAttributeValue(L"MinVersion", &minVersion));
+    std::wstring manifestMinVersion = minVersion.Get();
+
+    /// Major version
+    size_t start = 0;
+    size_t end = manifestMinVersion.find_first_of(L'.');
+    m_majorVersion = manifestMinVersion.substr(start, end - start);
+
+    /// Minor version
+    manifestMinVersion.replace(start, end - start + 1, L"");
+    end = manifestMinVersion.find_first_of(L'.');
+    m_minorVersion = manifestMinVersion.substr(start, end - start);
+
+    /// Build number
+    manifestMinVersion.replace(start, end - start + 1, L"");
+    end = manifestMinVersion.find_first_of(L'.');
+    m_buildNumber = manifestMinVersion.substr(start, end - start);
+
+    return S_OK;
+}
+
+bool ValidateTargetDeviceFamily::IsTargetDeviceFamilyNameCompatibleWithOS()
+{
+    if (isWindowsProductTypeDesktop()) /// Desktop OS
+    {
+        if (m_targetDeviceFamilyName == L"MsixCore.Server" || m_targetDeviceFamilyName == L"MsixCore.Desktop")

```

Comment:
>if (m_targetDeviceFamilyName == L"MsixCore.Server" || m_targetDeviceFamilyName == L"MsixCore.Desktop") [](start = 8, length = 102)

this is reverse-- this line for Deskktop OS should only take MsixCore.Desktop. The line for Server should allow both.

Reply:
Modified.

Reply:
update the comments too. 

---
In reply to: [277113289](https://github.com/Microsoft/msix-packaging/pull/91#discussion_r277113289) [](ancestors = 277113289)